Helm: add configuration options for kvstoremesh

giorio94 opened the pull request:
This PR is a followup of https://github.com/cilium/cilium/pull/26083, which extends the helm chart to allow configuring kvstoremesh.

```release-note
Extend the Helm chart to allow configuring kvstoremesh.
```

giorio94 committed:
kvstoremesh: add helm configuration

This commit extends the helm chart to allow configuring kvstoremesh. In
particular, the clustermesh-apiserver deployment is enriched with the
additional kvstoremesh sidecar container (when kvstoremesh is enabled),
appropriately mounting the secret containing the remote kvstore
configurations. Additionally, the configuration used by the agents is
modified to connect to the local kvstore instance (through the
corresponding service) instead of the remote ones.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
clustermesh: add service name to server certificate DNS names

Add the `clustermesh-apiserver.{namespace}.svc` DNS name to the
clustermesh-apiserver etcd certificate, so that the secure
connection can be successfully established when connecting to the
local etcd instance through the service, rather than to a remote one
through the corresponding LB/NodePort service or an external DNS name.
This is required in particular in the kvstoremesh case.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -2630,6 +2630,46 @@ clustermesh:
         #     cpu: 100m
         #     memory: 100Mi
 
+    kvstoremesh:
```
The `kvstoremesh` configuration options are currently placed at this level for consistency with `etcd`, but I'm wondering whether they should be moved one level up. WDYT?

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -2845,6 +2885,12 @@ clustermesh:
 
+      kvstoremesh:
+        # -- Enables exporting KVStoreMesh metrics in OpenMetrics format.
+        enabled: false
```
Can this be enabled by default? Since kvstoremesh isn't a host-network pod, it shouldn't cause any issues.

giorio94 replied:
I personally don't have a strong opinion here. ATM, metrics are disabled by default for consistency with the clustermesh-apiserver/etcd ones, but we could also switch those to be enabled by default in case.

giorio94 replied:
I've switched them to be enabled by default both for the clustermesh-apiserver and kvstoremesh.

tommyp1ckles commented on Documentation/observability/metrics.rst:
```diff
@@ -188,20 +188,24 @@ Prometheus namespace. Etcd metrics are exported under the ``etcd_`` Prometheus n
-``clustermesh.apiserver.metrics.enabled=true``.
-To enable metrics for the sidecar etcd instance, use
-``clustermesh.apiserver.metrics.etcd.enabled=true``.
+You can use the following settings to enable the metrics for the different
```
nit: settings -> values?

tommyp1ckles commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -2633,6 +2633,46 @@ clustermesh:
+      # -- KVStoreMesh image.
+      image:
+        override: ~
+        repository: "quay.io/cilium/kvstoremesh-ci"
```
this appears to be a CI image, is that intended?

giorio94 replied:
I've double-checked, and also the repositories for the other components point to CI images. It seems that they are sourced from [here](https://github.com/giorio94/cilium/blob/mio%2Fkvstoremesh-helm/install/kubernetes/Makefile.values#L12-L29) (and RELEASE is not set by default).

tommyp1ckles reviewed (approved):
looks good, just some minor questions/points.

giorio94 committed:
clustermesh: enable apiserver metrics by default in the Helm chart

The clustermesh-apiserver runs in pod network, hence there's no
shortcoming in enabling those metrics by default.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

zacharysarah commented on Documentation/observability/metrics.rst:
```diff
@@ -190,20 +190,24 @@ Prometheus namespace. Etcd metrics are exported under the ``etcd_`` Prometheus n
-To enable metrics for the sidecar etcd instance, use
-``clustermesh.apiserver.metrics.etcd.enabled=true``.
+You can use the following values to enable the metrics for the different
+Cluster Mesh API Server components:
```
Edits for clarity
```suggestion
You can enable the metrics for different
Cluster Mesh API Server components by setting the following values:
```

zacharysarah commented on Documentation/observability/metrics.rst:
```diff
@@ -190,20 +190,24 @@ Prometheus namespace. Etcd metrics are exported under the ``etcd_`` Prometheus n
      --set clustermesh.apiserver.metrics.etcd.enabled=true
 
-The ports can be configured via ``clustermesh.apiserver.metrics.port`` and
+The ports can be configured via ``clustermesh.apiserver.metrics.port``,
```
Use active voice, no Latin abbreviations
```suggestion
You can figure the ports by way of ``clustermesh.apiserver.metrics.port``,
```

zacharysarah reviewed (approved):
@giorio94 Good work. Approving to unblock with the understanding that changes are required prior to merge.
